### PR TITLE
A root and branch review of error messages.

### PR DESCRIPTION
**Pull request by bbakerman**
A root and branch review of error messages.

Now with toSpecification and extensions defined on them as well as an optional path.

I have made a series of tweaks so that all the messages will serialize without errors on Jackson and GSON but not to spec.

With the new toSpecification capability you can get spec valid errors

### CODE REVIEW

**Comment by bbakerman** [comment]
So while to get proper toSpecification errors you need to call toSpec on the ExecutionResult, some of these field naming changes just mean that without that its a "smidge" closer to working.

Better to be 75% correct than to be 5% correct say.  It is not 100% correct and cant be under GSON native say but it can be more correct than it otherwise would be.

For those that want 100% correct - they have `toSpecification`

File: docs/execution.rst [hunk 1]
```diff
@@ -107,6 +137,31 @@ behaviour.
...
+---------------------------
+
+The most common way to call graphql is over HTTP and to expect a JSON response back.  So you need to turn an
+`graphql.ExecutionResult` into a JSON payload.
+
+A common way to do that is use a JSON serialisation library like Jackson or GSON.  However exactly how they interpret the
+data result is particular to them.  For example `nulls` are important in graphql results and hence you must set up the json mappers
+to include them.
+
+To ensure you get a JSON result that confirms 100% to the graphql spec, you should called `toSpecification` on the result and then
```
Comment (andimarek) [inline docs/execution.rst]: typo: "you should called" -> "you should call"

File: src/main/java/graphql/GraphQLError.java [hunk 2]
```diff
@@ -22,19 +26,79 @@
...
+    }
+
 
     /**
      * This little helper allows GraphQlErrors to implement
-     * common things (hashcode/ equals) more easily
+     * common things (hashcode/ equals ) and to specification more easily
      */
     @SuppressWarnings("SimplifiableIfStatement")
     class Helper {
```
Comment (andimarek) [inline src/main/java/graphql/GraphQLError.java]: I don't like this helper inside an interface. Can we extract it to its own class/file?

Comment (bbakerman) [reply src/main/java/graphql/GraphQLError.java]: sure

File: docs/execution.rst (hunk 1 above)
Comment (bbakerman) [reply docs/execution.rst]: fix

**Pull request merged**